### PR TITLE
Creates draft documentation, makes torch_xla compatible with latest TensorFlow

### PR DESCRIPTION
**Pull request by mruberry**
- Adds docstrings to public functions and classes.
- Adds docs folder patterned off of torchvision's docs. Docs can be built using sphinx (sphinx-build -b html source build). 
- Creates draft torch_xla.rst document. See screen shot.
- Updates einsum handling to build with latest TensorFlow. See https://github.com/tensorflow/tensorflow/commit/8ebb1261481a0d05204b7204d33fbd092c3b4d0c.

To build the docs yourself use pip to install the requirements in requirements.txt, then run the above sphinx command in the docs directory.

Screen Shot 2019-09-30 at 7 14 34 AM

### CODE REVIEW

**Comment by mruberry** [comment]
@dlibenzi should I remove the changes to einsum? I'm not sure how often we update the tensorflow submodule.

File: docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+#
+# source_suffix = ['.rst', '.md']
+source_suffix = '.rst'
+
+# The master toctree document.
+master_doc = 'torch_xla'
+
+# General information about the project.
+project = 'TorchXLA'
+copyright = '<copyright>'
```
Comment (dlibenzi) [inline docs/source/conf.py]: In `LICENSE` we have `Copyright (c) 2018 Google Inc.` ... but there might be some talking there.

File: docs/source/torch_xla.rst [hunk 2]
```diff
@@ -0,0 +1,92 @@
+torch_xla
```
Comment (dlibenzi) [inline docs/source/torch_xla.rst]: We pervasively use Markdown for documentation.
Can we use that instead of RST?

File: docs/source/torch_xla.rst [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+===================================
+
+PyTorch runs on TPUs with the `torch_xla`_ package.
+
+XLA devices
+----------------------------------
+
+The XLA device type defines virtual 'xla' devices backed by real hardware.
+'xla:0' is your machine's CPU and is useful for testing.
```
Comment (dlibenzi) [inline docs/source/torch_xla.rst]: I would not expose that device is a given ordinal, as it might be subject to change.
Documenting properly `xm.xla_device()` and ask users to use that should suffice.

File: torch_xla/core/xla_model.py [hunk 4]
```diff
@@ -42,14 +50,29 @@ def get_xla_supported_devices(devkind=None, max_devices=None):
 
 
 def xrt_world_size(defval=1):
+  """
+  Returns the WORLD_SIZE environment variable, if defined, and defval
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py]: I would document a bit better here.
Ie, what WORLD_SIZE is..

File: torch_xla/core/xla_model.py [hunk 5]
```diff
@@ -42,14 +50,29 @@ def get_xla_supported_devices(devkind=None, max_devices=None):
...
+  otherwise.
+  """
+
   return xu.getenv_as(xenv.WORLD_SIZE, int, defval=defval)
 
 
 def get_ordinal(defval=0):
+  """
+  Returns the ORDINAL environment variable, if defined, and defval
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py]: Same here for ORDINAL.

File: torch_xla/core/xla_model.py [hunk 6]
```diff
@@ -42,14 +50,29 @@ def get_xla_supported_devices(devkind=None, max_devices=None):
...
+  otherwise.
+  """
+
   return xu.getenv_as(xenv.ORDINAL, int, defval=defval)
 
 
 def is_master_ordinal():
+  """
+  Returns true if this process or thread is the 'master' process or thread,
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py]: Same here ... mentioning that the `master` ordinal is ordinal 0 would make sense IMHO.

File: torch_xla/core/xla_model.py [hunk 7]
```diff
@@ -92,6 +131,12 @@ def xla_real_devices(devices):
 
 
 def xla_replication_devices(local_devices):
+  """
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py]: I would not document this, especially with these very simple comments.
Leave them out as those are internal.

File: torch_xla/core/xla_model.py [hunk 8]
```diff
@@ -120,6 +165,11 @@ def xla_replication_devices(local_devices):
 
 
 def set_replication(device, devices):
+  """
+    Sets the current thread's device to device and sets replication devices
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py]: Ditto as for replciation_devices() above.

File: torch_xla/csrc/aten_xla_type.cpp [hunk 9]
```diff
@@ -1015,9 +1015,17 @@ at::Tensor& AtenXlaType::dropout_(at::Tensor& self, double p, bool train) {
 }
 
 at::Tensor AtenXlaType::einsum(std::string equation, at::TensorList tensors) {
-  if (tensors.size() != 2 || !ir::ops::Einsum::SupportsEquation(equation)) {
+  if (tensors.size() != 2) {
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: What is this change?
Does not seem to matter with documentation.
I'd move to a different PR and explain better in the comment ...
Ditto for all the related C++ files in this PR.

File: torch_xla/distributed/data_parallel.py [hunk 10]
```diff
@@ -64,6 +64,9 @@ def next(self):
 
 
 class ParallelLoader(object):
+  """
```
Comment (dlibenzi) [inline torch_xla/distributed/data_parallel.py]: I wonder ... do these comments really help?

Comment (mruberry) [reply torch_xla/distributed/data_parallel.py]: No. They're placeholders. It's more important to identify the functions and classes that need docstrings and ensure they're built properly at this point.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 9 above)
Comment (mruberry) [reply torch_xla/csrc/aten_xla_type.cpp]: I couldn't build without this update. I'll remove it from this PR.

File: torch_xla/core/xla_model.py (hunk 8 above)
Comment (mruberry) [reply torch_xla/core/xla_model.py]: Sounds good. I updated our spreadsheet to reflect that these are undocumented. Should we change private functions' names to start with an underscore?

File: torch_xla/core/xla_model.py (hunk 7 above)
Comment (mruberry) [reply torch_xla/core/xla_model.py]: _xla_replication_devices?

File: docs/source/torch_xla.rst (hunk 2 above)
Comment (mruberry) [reply docs/source/torch_xla.rst]: I'll check with our docs team.

File: torch_xla/core/xla_model.py (hunk 7 above)
Comment (dlibenzi) [reply torch_xla/core/xla_model.py]: Not that private :smile: ... we use it internally.
I'd leave the name w/out '_' as usually by our convention, that is file private.

File: torch_xla/csrc/aten_xla_type.cpp (hunk 9 above)
Comment (dlibenzi) [reply torch_xla/csrc/aten_xla_type.cpp]: Did you update TF?
I can see the signature could have changed on the TF side, but you should be using the commit ID which comes with our submodule.

File: torch_xla/distributed/data_parallel.py (hunk 10 above)
Comment (dlibenzi) [reply torch_xla/distributed/data_parallel.py]: I am OK with these, but they need better comments.
